x/upgrade/doc.go

```diff
@@ -0,0 +1,56 @@
+/*
+Package upgrade provides a Cosmos SDK module that can be used for smoothly upgrading a live Cosmos chain to a
+new software version. It accomplishes this by providing a BeginBlocker hook that prevents the blockchain state
+machine from proceeding once a pre-defined upgrade block time or height has been reached. The module does not prescribe
+anything regarding how governance decides to do an upgrade, but just the mechanism for coordinating the upgrade safely.
+Without software support for upgrades, upgrading a live chain is risky because all of the validators need to pause
+their state machines at exactly the same point in the process. If this is not done correctly, there can be state
+inconsistencies which are hard to recover from.
+
+Integrating With An App
+
+Setup an upgrade Keeper for the app and then define a BeginBlocker that calls the upgrade
+keeper's BeginBlocker method:
+    func (app *myApp) BeginBlocker(ctx sdk.Context, req abci.RequestBeginBlock) abci.ResponseBeginBlock {
+    	app.upgradeKeeper.BeginBlocker(ctx, req)
+    	return abci.ResponseBeginBlock{}
+    }
+
+The app must then integrate the upgrade keeper with its governance module as appropriate. The governance module
+should call ScheduleUpgrade to schedule an upgrade and ClearUpgradePlan to cancel a pending upgrade.
+
+Performing Upgrades
```

Comment:
This is an interesting idea. And wonderful documentation. Both these package-level comments, as well as all the comments on types.

For halting, this works well. However, I think we need a more complete strategy for upgrades. I will expand on this o the issue. But yeah, this seems a nice first step.

Reply:
Does this file need to get updated with the changes that have been made?

Reply:
we need to eventually transition or duplicate part of this godoc to a new `docs/` guideline for how to upgrade live chain. cc: @gamarin2 @hschoenburg